### PR TITLE
Fix font em size

### LINKED ISSUE #160: Validate `emSize` in `GdipCreateFont`
In `test_createFont` in `testfont.c`

```cpp
	// FIXME: there are several places that indirectly call GdipCreateFont in
	// libgdiplus but allow negative em sizes on GDI+ because the GDI+
	// implementation does not call GdipCreateFont. An example is GdipAddPathString.
	// A fix for this will need to carefully account for places where this
	// API is called.
#if defined(USE_WINDOWS_LIBGDIPLUS)
	status = GdipCreateFont (family, -1, 10, UnitPixel, &font);
	assertEqualInt (status, InvalidParameter);

	status = GdipCreateFont (family, 0, 10, UnitPixel, &font);
	assertEqualInt (status, InvalidParameter);
#endif
```

### LINKED ISSUE #321: GdipAddPathString does not accept -1 as a length for the string parameter
A value of -1 means "find the string length yourself"

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #321
Fixes #160 

~Relies on #227~

### CODE REVIEW

File: src/graphics-path.c [hunk 1]
```diff
@@ -1187,7 +1199,7 @@ GdipAddPathString (GpPath *path, GDIPCONST WCHAR *string, int length,
 		return OutOfMemory;
 	}
 
-	status = GdipCreateFont (family, emSize, style, UnitPixel, &font);
+	status = GdipCreateFont (family, 10, style, UnitPixel, &font);
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/graphics-path.c]: One possible approach is to create the font in its native size (GdipGetEmHeight) and then do a lot of scaling throughout the function. I don't understand all the nuances of doing that though.

**Comment by hughbe** [comment]
I've fixed the scaling bug and tested things against GDI+ and verified the expected behaviour with the cairo backend.

Unfortunately the pango backend crashes (even before this PR): https://github.com/mono/libgdiplus/issues/325

**Comment by filipnavara** [comment]
The Pango code doesn't work because of this line:
`layout = gdip_pango_setup_layout ((GpGraphics *)cr, string, length, font, layoutRect, &box, &box_offset, format, NULL);`

'cr' is not castable to `GpGraphics *`. The easiest way to fix it is probably to create a GDI+ bitmap instead of cairo surface and then use `graphics->cr` whereever `cr` is used now.

**Comment by filipnavara** [comment]
Pango fixed in https://github.com/mono/libgdiplus/pull/365

**Comment by hughbe** [comment]
Something still isn't right, however. I ran the tests against cairo and got the expected result when dumping the following (L107-114)

```

  GdipCreatePath (FillModeAlternate, &path);

  status = GdipAddPathString (path, string, 5, family, 0, 72, &layoutRect, format);
  assertEqualInt (status, Ok);

  dumpPath (path);
  GdipDeletePath (path);
```

However, with pango, I got a black page

**Comment by filipnavara** [comment]
Alright, will look into it.

**Comment by filipnavara** [comment]
It's a combination of two different problems:
1) Pango cares about the layout rectangle, so if you render the font at the native height it exceeds the rectangle and gets clipped.
2) gdip_pango_setup_layout applies the clipping from the Cairo surface. In this case the surface has 1x1 size, so it's obviously not a good idea.

**Comment by marek-safar** [comment]
@monojenkins build

**Pull request merged**